src/main/java/com/gooddata/collections/MultiPageableList.java

```diff
@@ -0,0 +1,233 @@
+        return delegate.containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(final Collection<? extends T> c) {
+        return delegate.addAll(c);
+    }
+
+    @Override
+    public boolean addAll(final int index, final Collection<? extends T> c) {
+        return delegate.addAll(index, c);
+    }
+
+    @Override
+    public boolean removeAll(final Collection<?> c) {
+        return delegate.removeAll(c);
+    }
+
+    @Override
+    public boolean retainAll(final Collection<?> c) {
+        return delegate.retainAll(c);
+    }
+
+    @Override
+    public void clear() {
+        delegate.clear();
+    }
+
+    @Override
+    public T get(final int index) {
+        return delegate.get(index);
+    }
+
+    @Override
+    public T set(final int index, final T element) {
+        return delegate.set(index, element);
+    }
+
+    @Override
+    public void add(final int index, final T element) {
+        delegate.add(index, element);
+    }
+
+    @Override
+    public T remove(final int index) {
+        return delegate.remove(index);
+    }
+
+    @Override
+    public int indexOf(final Object o) {
+        return delegate.indexOf(o);
+    }
+
+    @Override
+    public int lastIndexOf(final Object o) {
+        return delegate.lastIndexOf(o);
+    }
+
+    @Override
+    public ListIterator<T> listIterator() {
+        return delegate.listIterator();
+    }
+
+    @Override
+    public ListIterator<T> listIterator(final int index) {
+        return delegate.listIterator(index);
+    }
+
+    @Override
+    public List<T> subList(final int fromIndex, final int toIndex) {
+        return delegate.subList(fromIndex, toIndex);
+    }
+
+    private static final class PageIterator<T> implements Iterator<T> {
+
+        private PageableList<T> currentPage;
+        private Iterator<T> pageItemsIterator;
+        private final Function<Page, PageableList<T>> pageProvider;
+
+        PageIterator(final PageableList<T> currentPage, final Function<Page, PageableList<T>> pageProvider) {
+            this.pageProvider = pageProvider;
+            this.currentPage = currentPage;
+            this.pageItemsIterator = currentPage.getItems().iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (pageItemsIterator.hasNext()) {
+                return true;
+            }
+
+            if (currentPage.getNextPage() == null) {
+                return false;
+            }
+
+            final PageableList<T> next = pageProvider.apply(currentPage.getNextPage());
+            if (currentPage.getNextPage().equals(next.getNextPage())) {
+                throw new IllegalArgumentException("page provider does not iterate properly, returns the same page");
+            }
+
+            if (next.isEmpty() && next.getNextPage() != null) {
+                throw new IllegalStateException("page has no results, yet claims there is next page");
+            }
+
+            currentPage = next;
```

Comment:
We won't know if next page has some further items unless we load it first, right? Or is it safe to assume that if there is a link to the next page, it will surely have at least one item?